sweep/sweeper.go

```diff
@@ -30,13 +63,671 @@ type UtxoSweeperConfig struct {
+	}
+
+	// Inputs are no longer pending after result has been sent.
+	delete(s.pendingInputs, *outpoint)
+}
+
+// sweepAll goes through all pending inputs and constructs sweep transactions,
+// each up to the configured maximum number of inputs. Negative yield inputs are
+// skipped. Transactions with an output below the dust limit are not published.
+// Those inputs remain pending and will be bundled with future inputs if
+// possible.
+//
+// Returns a boolean indicating whether there are sweepable inputs remaining.
+// This can happen when fullOnly is true and there are less sweepable inputs
+// then the configured maximum number per tx.
+func (s *UtxoSweeper) sweepAll(fullOnly bool) (bool, error) {
+	// Retrieve fee estimate for input filtering and final tx fee
+	// calculation.
+	satPerKW, err := s.cfg.Estimator.EstimateFeePerKW(
+		s.cfg.SweepTxConfTarget)
+	if err != nil {
+		return false, err
+	}
+
+	// Retrieve current height to properly set tx locktime.
+	_, currentHeight, err := s.cfg.ChainIO.GetBestBlock()
+	if err != nil {
+		return false, err
+	}
+
+	// Filter for inputs that need to be swept.
+	var sweepableInputs []*pendingInput
+	for _, input := range s.pendingInputs {
+		// Skip inputs that are already part of a sweep tx.
+		if input.state != stateNew {
+			continue
+		}
+
+		sweepableInputs = append(sweepableInputs, input)
+	}
+
+	// Sort input by yield. We will start constructing sweep transactions
+	// starting with the highest yield inputs. This is to prevent the
+	// construction of a sweep tx with an output below the dust limit,
+	// causing the sweep process to stop, while there are still higher value
+	// inputs sweepable. It also allows us to stop evaluating more inputs
+	// when the first input in this ordering is encountered with a negative
+	// yield.
+	//
+	// Yield is calculated as the difference between value and added fee for
+	// this input. The fee calculation excludes fee components that are
+	// common to all inputs, as those wouldn't influence the order. The
+	// single component that is differentiating is witness size.
+	//
+	// For witness size, a worst case upper limit is taken. The actual size
+	// depends on the signature length, which is not known yet at this
+	// point.
+	yield := func(i *pendingInput) int64 {
+		return i.input.SignDesc().Output.Value -
+			int64(satPerKW.FeeForWeight(
+				int64(i.witnessSizeUpperBound)))
+	}
+
+	sort.Slice(sweepableInputs, func(i, j int) bool {
+		return yield(sweepableInputs[i]) > yield(sweepableInputs[j])
+	})
+
+	// Select blocks of inputs up to the configured maximum number.
+	stopSweep := false
+	for len(sweepableInputs) > 0 && !stopSweep {
+		var weightEstimate lnwallet.TxWeightEstimator
+
+		// Add the sweep tx output to the weight estimate.
+		weightEstimate.AddP2WKHOutput()
+
+		var inputList []Input
+		var total, outputValue int64
+		for len(sweepableInputs) > 0 {
+			input := sweepableInputs[0]
+			sweepableInputs = sweepableInputs[1:]
```

Comment:
Should set `sweepableInputs[0] = nil` to avoid a GC leak: https://github.com/golang/go/wiki/SliceTricks#delete-without-preserving-order
> NOTE If the type of the element is a pointer or a struct with pointer fields, which need to be garbage collected, the above implementations of Cut and Delete have a potential memory leak problem: some elements with values are still referenced by slice a and thus can not be collected. The following code can fix this problem:
```go
copy(a[i:], a[i+1:])
a[len(a)-1] = nil // or the zero value of T
a = a[:len(a)-1]
```

Reply:
Interesting fact. But I don't think it is necessary to do here, because the slice is a temporary list and goes out of scope pretty quickly. It isn't stored in the `sweeper` struct.